File: src/Rashtan.AuditStory.Payment/PaymentGizmo.fs
namespace Rashtan.AuditStory.Payment

open Microsoft.Extensions.Configuration
open Braintree;
open Rashtan.AuditStory.Dto

type IPaymentGizmo = 
    abstract member GenerateNonceAsync: unit -> Async<string>
    abstract member SaleAsync: payment: PaymentToProcess -> Async<Result<PaymentProcessed, string>>

type PaymentGizmo(configuration: IConfiguration) =
    
    let gateway =       
        let payment = configuration.GetSection("Payment");
        let gateway = BraintreeGateway()
        gateway.Environment <- Braintree.Environment.ParseEnvironment(payment.["Environment"])
        gateway.MerchantId <- payment.["MerchantId"]
        gateway.PublicKey <- payment.["PublicKey"]
        gateway.PrivateKey <- payment.["PrivateKey"]
        gateway

    interface IPaymentGizmo with
        member __.GenerateNonceAsync() = async {
            return! gateway.ClientToken.GenerateAsync() |> Async.AwaitTask
        }

        member __.SaleAsync (payment: PaymentToProcess) = async {
            let request = 
                let request = TransactionRequest()
                request.Amount <- payment.Amount
                request.PaymentMethodNonce <- payment.Nonce
                request.Options <- TransactionOptionsRequest()
                request.Options.SubmitForSettlement <- System.Nullable(true)
                request
            try 
                let! result = gateway.Transaction.SaleAsync(request) |> Async.AwaitTask
            
                if result.IsSuccess() then
                    let paymentProcessed = {
                        TransactionId = result.Target.AuthorizedTransactionId
                        Amount = payment.Amount
                        Currency = "USD"
                        Method = result.Target.Channel
                        PayedAt = System.DateTime.Today.ToUniversalTime()
                        PayedUntil = System.DateTime.Today.AddDays(payment.Length |> float).ToUniversalTime()
                    }

                    return Ok paymentProcessed;
                else
                    return Error result.Message;
            with
                | ex -> return Error ex.Message
            }

File: src/Rashtan.AuditStory.PaymentGizmo/PaymentGizmo.cs
using Braintree;
using Microsoft.Extensions.Configuration;
using Microsoft.FSharp.Core;
using System;
using System.Threading.Tasks;
using static Rashtan.AuditStory.Dto.Payment;

namespace Rashtan.AuditStory.PaymentGizmo
{
    internal class PaymentGizmo : IPaymentGizmo
    {
        private BraintreeGateway Gateway { get; }

        public PaymentGizmo(IConfiguration configuration)
        {
            var payment = configuration.GetSection("Payment");

            Gateway = new BraintreeGateway
            {
                Environment = Braintree.Environment.ParseEnvironment(payment[nameof(BraintreeGateway.Environment)]),
                MerchantId = payment[nameof(BraintreeGateway.MerchantId)],
                PublicKey = payment[nameof(BraintreeGateway.PublicKey)],
                PrivateKey = payment[nameof(BraintreeGateway.PrivateKey)]
            };
        }

        public async Task<string> GenerateNonceAsync() => await Gateway.ClientToken.GenerateAsync();

        public async Task<FSharpResult<PaymentProcessed, string>> SaleAsync(PaymentToProcess payment)
        {
            var request = new TransactionRequest
            {
                Amount = payment.Amount,
                PaymentMethodNonce = payment.Nonce,
                Options = new TransactionOptionsRequest
                {
                    SubmitForSettlement = true
                }
            };

            try
            {
                var result = await Gateway.Transaction.SaleAsync(request);
                if (result.IsSuccess())
                {
                    var paymentProcessed = new PaymentProcessed
                    {
                        TransactionId = result.Target.AuthorizedTransactionId,
                        Amount = payment.Amount,
                        PayedUntil = DateTime.Today.AddDays(payment.Length)
                    };

                    return FSharpResult<PaymentProcessed, string>.NewOk(paymentProcessed);
                }
                else
                {
                    return FSharpResult<PaymentProcessed, string>.NewError(result.Message);
                }
            }
            catch (Exception e)
            {   
                return FSharpResult<PaymentProcessed, string>.NewError(e.Message);
            }            
        }
    }
}
